k-distribution/src/main/scripts/lib/pyk/pyk/cterm.py

```diff
@@ -0,0 +1,105 @@
+
+
+@dataclass(frozen=True)
+class CTerm:
+    config: KInner  # TODO Optional?
+    constraints: Tuple[KInner, ...]
+
+    def __init__(self, term: KInner) -> None:
+        config, constraint = splitConfigAndConstraints(term)
+        constraints = tuple(flattenLabel('#And', constraint))
+        object.__setattr__(self, 'config', config)
+        object.__setattr__(self, 'constraints', constraints)
+
+    def __iter__(self):
+        return chain([self.config], self.constraints)
+
+    @cached_property
+    def term(self) -> KInner:
+        return mlAnd(self)
+
+    @property
+    def hash(self) -> str:
+        return self.term.hash
+
+    @staticmethod
+    def from_rule(rule: KRuleLike) -> Tuple['CTerm', 'CTerm']:
+        body = drop_var_prefixes(rule.body)
+        lhs = extract_lhs(body)
+        rhs = extract_rhs(body)
+        requires = drop_var_prefixes(rule.requires)
+        ensures = drop_var_prefixes(rule.ensures)
+        return CTerm(mlAnd([lhs, requires])), CTerm(mlAnd([rhs, ensures]))
+
+    def to_rule(self, other: 'CTerm', path_constraints: Iterable[KInner] = [], *, claim=False) -> KRuleLike:
+        lhs_constraints = set(chain(self.constraints, path_constraints))
+        rhs_constraints = set(c for c in other.constraints if c not in lhs_constraints)
+
+        body = pushDownRewrites(KRewrite(self.config, other.config))
+        requires = simplifyBool(unsafeMlPredToBool(mlAnd(lhs_constraints)))
+        ensures = simplifyBool(unsafeMlPredToBool(mlAnd(rhs_constraints)))
+
+        var_count = count_vars(body) + count_vars(requires) + count_vars(ensures)
+        rhs_var_count = count_rhs_vars(body) + count_vars(ensures)
+
+        singletons = {k for k, v in var_count.items() if v == 1}
+        existentials = {k for k, v in rhs_var_count.items() if var_count[k] <= v}
+
+        subst_items = ((k, '_' + k) for k in singletons)
+        subst_items = ((k, ('?' if k in existentials else '') + v) for k, v in subst_items)
+        subst_dict = {k: KVariable(v) for k, v in subst_items}
+        subst = Subst(subst_dict)
+
+        body = subst(body)
+        requires = subst(requires)
+        ensures = subst(ensures)
+
+        if claim:
+            return KClaim(body=body, requires=requires, ensures=ensures)
+
+        return KRule(body=body, requires=requires, ensures=ensures)
+
+    def match(self, pattern: 'CTerm') -> Optional[Tuple[Subst, Set[KInner], Set[KInner]]]:
+        subst = match(pattern=pattern.config, kast=self.config)
+
+        if subst is None:
+            return None
+
+        subst = Subst(subst)
+
+        antecedents = set(flattenLabel('_andBool_', simplifyBool(unsafeMlPredToBool(mlAnd(self.constraints)))))
+        consequents = {
+            conjunct
+            for conjunct in flattenLabel('_andBool_', simplifyBool(unsafeMlPredToBool(subst(mlAnd(pattern.constraints)))))
+            if conjunct not in antecedents
+        }
+
+        return subst, antecedents, consequents
```

Comment:
This seems backward to me. I would do `pattern.match(term)` instead of `term.match(pattern)`. I guess I'm not sure actually though. I would normally say that "`pattern` _matches_ `term`", or "`term` _is matched by_ `pattern`".

I guess it probably doesn't matter too much.

But also, we should not be calling `simplifyBool` or `unsafeMlPredToBool` here I don't think.... it seems to assume too much about the structure. I think this same implementation can happen without calling those functions, the only thing of note that `simplifyBool` is probably doing is (i) removing any instances of `#Top`, and (ii) deduplicating clauses. Both of which should be maintained as invariants of the `CTerm` class anyway.

Reply:
If we do `pattern.match(term)`, we can also have `pattern.substitute(term)`, and should have the property:

```
match = pattern.match(term)
if match:
    new_term = pattern.substitute(match)
    assertEqual(term, new_term)
```

Reply:
Based on these two Google searches over stackoverflow.com, both forms are used (I searched for "string" as that is the most common form of matching):

* ["pattern matches a string"](https://www.google.com/search?q=%22pattern+matches+a+string%22+site%3Astackoverflow.com)
* ["string matches a pattern"](https://www.google.com/search?q=%22string+matches+a+pattern%22+site%3Astackoverflow.com)

Sometimes even in the same sentence: https://stackoverflow.com/a/45632286 :slightly_smiling_face:

I prefer `term.match(pattern)` for two reasons:

* It aligns with the order formulas in an implication (or order of sets of models in a set inclusion), so something like this reads more naturally: `if term.match(pattern): ...`.
* `def match(pattern)` makes it obvious that the term to be passed as an argument is the pattern.

The property you identified is nice though: `pattern.substitute(pattern.match(term)) == term`. (Right now we have `term.match(pattern)(pattern) == term`.)

We can also consider which one of the two functions is more likely to be passed to a higher order function as argument, for example:

* `map(term.match, [pattern1, pattern2, pattern3])` - match a term over multiple patterns, which is analogous to FP pattern matching (e.g. `term match { case pattern1 => ... }` in Scala)
* `map(pattern.match, [term1, term2, term2])` - match a pattern over multiple terms, which is some sort of classification of terms w.r.t. pattern.

Reply:
> simplifyBool is probably doing is (i) removing any instances of #Top, and (ii) deduplicating clauses. Both of which should be maintained as invariants of the CTerm class anyway

Good point about maintaining the invariant. Though this function also does rewrites such as: `(true ==K (N >=Int 0)) => (N >=Int 0)`

W.rt. `unsafeMlPredToBool`: is there an other way to ensure that no ML connectives leak into `requires`-`ensures`? Or can we assume that the only ML connectives that occur are `#Top` and `#And`?

Reply:
Let's make sure we have tests of `pattern.substitute(term.match(pattern)) == term`.